target_postgres/stream_tracker.py

```diff
@@ -21,22 +21,29 @@ def __init__(self, target, emit_states):
         self.emit_states = emit_states
 
         self.streams = {}
+
+        # dict of {'<stream_name>': number}, where the number is the message counter of the most recently received record for that stream. Will contain a value for all registered streams.
         self.stream_add_watermarks = {}
+
+        # dict of {'<stream_name>': number}, where the number is the message counter of the most recently flushed record for that stream. Will contain a value for all registered streams.
         self.stream_flush_watermarks = {}
 
+        self.streams_added_to = set()  # list of stream names which have seen records
         self.state_queue = deque()  # contains dicts of {'state': <state blob>, 'watermark': number}
         self.message_counter = 0
         self.last_emitted_state = None
 
     def register_stream(self, stream, buffered_stream):
         self.streams[stream] = buffered_stream
-        self.stream_add_watermarks[stream] = 0
         self.stream_flush_watermarks[stream] = 0
```

Comment:
I think if we actually delay adding the `stream_flush_watermarks` here _as well_, we can avoid entirely the usage of `streams_added_to`.

This would only require need `_flush_stream`'s _original_ logic to be modified as:

```py
self.stream_flush_watermarks[stream] = self.stream_add_watermarks.get(stream, 0)
```

Then `_emit_safe_queued_states`'s _original_ logic wouldn't need to change at all right?

ie, if we only set `stream_flush_watermarks` when something is flushed (which should only typically happen when the `buffer_full` anyways, but I digress) and then `stream_add_watermarks` when something is added we should be in the clear...?

Reply:
I actually had it that way when I first put this patch together where `stream_flush_watermarks` had double meaning and captured both the notion of if a stream had been added to and when that stream had been last flushed. But, I felt that because the code was already pretty complicated (and bug prone) that it would be wise to not layer on a second, only somewhat related meaning to the idea of flushing, and so I added the new set to be extra clear. I think it's a little more verbose and will take a couple extra bytes of memory but is a bit clearer for the next poor soul that comes along to touch this code.